sdk/metric/processor/basic/basic.go

```diff
@@ -117,11 +117,12 @@ type (
 )
 
 var _ export.Processor = &Processor{}
+var _ export.Checkpointer = &Processor{}
 var _ export.CheckpointSet = &state{}
 var ErrInconsistentState = fmt.Errorf("inconsistent processor state")
 var ErrInvalidExporterKind = fmt.Errorf("invalid exporter kind")
 
-// New returns a basic Processor using the provided
+// New returns a basic Processor, also a Checkpointer, using the provided
```

Comment:
Reflecting on this change, I think the biggest problem is the creation of a new name when there is only one type involved. This arises because:

- the Accumulator only needs to call Process(), has no need for CheckpointSet() API
- the Accumulator _could_ call StartCollection() and FinishCollection(), but this would limit our ability to combine outputs from multiple Accumulators
- a Controller needs to consider locking and synchronization (and potentially multiple Accumulators) before it calls the Accumulator to Collect()

Therefore, the Controller wants to see a Checkpointer interface, and an Accumulator only needs a Processor interface (using the names in this PR). OTOH, these are just different APIs to a "Processor". Some other names might be: "AccumulationProcessor" and "CheckpointProcessor". Controllers need CheckpointProcessors, Accumulators need AccumulationProcessors. I will be happy to reconsider the names.

> goal is to have a reducer act as a Processor which will wrap a Checkpointer which wraps an embedded Processor

I was not thinking of a checkpointer as having an embedded processor. They're really the same, just two interfaces, that's why I think a naming change might clear this up.

> Does this need to be rethought as an idiomatic Go processing pipeline (using channels or a dedicated processing stack)?

I don't think so. For channels, it would create complexity. Today the Controller oversees the Accumulator and Processor and locking. If you had channels, the transactional nature of Start/Collect/Finish/CheckpointSet would be harder to implement. I think I see this more as a processing stack already: in the next PR, a Reducer Processor will be added to support dropping labels. The difference between the Reducer processor and the Basic processor is that the Basic processor is capable of being the end of the line, whereas the Reducer is only useful as an intermediate stage. Still, the Accumulator sees both the Reducer and the basic Processor as just processors.